engine/server/engine/server/websocket_api_handler.go

```diff
@@ -300,3 +340,95 @@ func streamServiceLogsWithHTTP(ctx echo.Context, streamer streaming.ServiceLogSt
 		}
 	}
 }
+
+type WebsocketPump[T interface{}] struct {
+	websocket  *websocket.Conn
+	inputChan  chan *T
+	infoChan   chan *api_type.ResponseInfo
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+func NewWebsocketPump[T interface{}](ctx echo.Context, cors cors.Cors) (*WebsocketPump[T], error) {
+	// nolint: exhaustruct
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  wsReadBufferSize,
+		WriteBufferSize: wsWriteBufferSize,
+		CheckOrigin:     cors.OriginAllowed,
+	}
+
+	conn, err := upgrader.Upgrade(ctx.Response(), ctx.Request(), nil)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Failed to upgrade http connection to websocket")
+	}
+
+	ctxWithCancel, cancelFunc := context.WithCancel(context.Background())
+
+	return &WebsocketPump[T]{
+		websocket:  conn,
+		inputChan:  make(chan *T),
+		ctx:        ctxWithCancel,
+		cancelFunc: cancelFunc,
+	}, nil
+}
+
+func (pump WebsocketPump[T]) startPumping() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		pump.websocket.Close()
+	}()
+
+	logrus.WithFields(logrus.Fields{
+		"pongWait":       pongWait,
+		"pingPeriod":     pingPeriod,
+		"maxMessageSize": maxMessageSize,
+	}).Debug("Started keep alive process for websocket connection.")
+
+	pump.websocket.SetReadLimit(maxMessageSize)
+	if err := pump.websocket.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		logrus.WithError(err).Error("Failed to set Pong wait time")
+		return
+	}
+	// nolint:errcheck
+	pump.websocket.SetPongHandler(func(string) error { return pump.websocket.SetReadDeadline(time.Now().Add(pongWait)) })
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := pump.websocket.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+				logrus.Debug("Websocket connection is likely closed, exiting keep alive process")
+				return
+			}
+			if err := pump.websocket.WriteMessage(websocket.PingMessage, nil); err != nil {
+				logrus.Debug("Websocket connection is likely closed, exiting keep alive process")
+				return
+			}
+		case msg := <-pump.inputChan:
+			if err := pump.websocket.WriteJSON(msg); err != nil {
+				logrus.WithError(stacktrace.Propagate(err, "Failed to send value of type `%T` via websocket", msg)).Errorf("Failed to write message to websocket, closing it.")
+				return
+			}
+		case msg := <-pump.infoChan:
+			if err := pump.websocket.WriteJSON(msg); err != nil {
+				logrus.WithError(stacktrace.Propagate(err, "Failed to send value of type `%T` via websocket", msg)).Errorf("Failed to write message to websocket, closing it.")
+				return
+			}
+		case <-pump.ctx.Done():
+			logrus.Debug("Websocket pumper has been asked to close, closing it.")
+			return
+		}
+	}
+}
+
+func (pump *WebsocketPump[T]) pumpResponseInfo(msg *api_type.ResponseInfo) {
+	pump.infoChan <- msg
+}
+
+func (pump *WebsocketPump[T]) pumpMessage(msg *T) {
+	pump.inputChan <- msg
+}
+
+func (pump *WebsocketPump[T]) Close() {
+	pump.cancelFunc()
```

Comment:
I think you can close the web socket here too and the channels, or is it necessary to let them open?

Reply:
and the ticker?

Reply:
If I do it here, we'll be closing it twice and it'll throw an error. The goroutine `StartPumping()` will always close it on exit (`defer`).

Reply:
@lostbean makes sense, so the close channels could be included in the defer on exit? it is suggested to always close the channels when the writer ends the task

Reply:
yes, I forgot to close the go channels!

Reply:
done!